lib/std/padded_array.zig

```diff
@@ -0,0 +1,118 @@
+const std = @import("std.zig");
+const assert = std.debug.assert;
+const mem = std.mem;
+const testing = std.testing;
+
+/// A structure that implements an array where the length is terminated by a sentinel value.
+/// This is useful for wrapping external string types or arrays which have an implicit length.
+/// The array contains a maximum of `length` elements of type `T`. The length is determined by
+/// the first appearance of the `sentinel` value.
+pub fn PaddedArray(comptime T: type, comptime length: usize, comptime sentinel: T) type {
+    return PaddedArrayExtra(T, length, sentinel, .Auto);
+}
+
+/// A structure that implements an array where the length is terminated by a sentinel value.
+/// This is useful for wrapping external string types or arrays which have an implicit length.
+/// The array contains a maximum of `length` elements of type `T`. The length is determined by
+/// the first appearance of the `sentinel` value.
+///
+/// This variant can define the structure layout, so it can be used with external or packed variants.
+pub fn PaddedArrayExtra(comptime T: type, comptime length: usize, comptime sentinel: T, comptime layout: std.builtin.Type.ContainerLayout) type {
+    return switch (layout) {
+        .Auto => struct {
+            pub usingnamespace PaddedArrayImplementation(@This(), T, length, sentinel);
+            items: [length]T = [1]T{sentinel} ** length,
+        },
+        .Packed => packed struct {
+            pub usingnamespace PaddedArrayImplementation(@This(), T, length, sentinel);
+            items: [length]T = [1]T{sentinel} ** length,
+        },
+        .Extern => extern struct {
+            pub usingnamespace PaddedArrayImplementation(@This(), T, length, sentinel);
+            items: [length]T = [1]T{sentinel} ** length,
```

Comment:
Is there any reason to not make this `items: [length:sentinel]T`?

Reply:
Yes, because you have no guarantee that the sentinel will be there. A C program might give you 32 non-NUL characters in a `char[32]`

Reply:
Would that be an issue though? The only places I've seen the sentinel be safety check is in the sentinel slicing operation (`[a..b:s]`).

Reply:
Yes, because the sentinel is guaranteed by design. Sentinelled arrays are afaik just a hack to allow string literals to be pointers to arrays and both coerce to a C string pointer.

It's also often done that these arrays would have length 32 fully used. Enforcing the sentinel there would prevent the modeling of this.